### ISSUE TITLE
Deploying SmolVLA with a simulator

### ISSUE THREAD

**Issue by aditya1709** [huggingface/lerobot#1923]
Has anyone been able to deploy the SmolVLA model to control say the SO-100 on a simulator like IsaacSim? 
Even if the fine-tuning reliably converges the observed performance on the simulator seems erratic. Do we apply the predicted actions from SmolVLA directly into the Articulation controller as positions?

**Comment by jadechoghari** [comment]
Hi! Thanks for the question 👋

SmolVLA does not use delta actions in its default setup, it outputs absolute joint position targets (normalized during training with MEAN_STD, then unnormalized at inference). in LIBERO for example, these predicted actions are applied via a position controller, which matches the action space the model was trained on.

So if you’re deploying to IsaacSim, you’ll want to:
- unnormalize the actions before applying them.
- feed them into your articulation controller as joint position targets (not raw deltas)
- you can make sure/ base yourself that the control interface matches the LIBERO convention, otherwise the behavior may look erratic

At the end it depends on what data you want it trained on.
And tip: make sure the observation key names matches what the policy expects! you can check libero docs here: https://huggingface.co/docs/lerobot/en/libero

**Comment by jadechoghari** [comment]
This issue has been addressed, so I’ll close it for now.
Please feel free to reopen if needed.

**Comment by MyLovelyAxe** [comment]
> Hi! Thanks for the question 👋
> 
> SmolVLA does not use delta actions in its default setup, it outputs absolute joint position targets (normalized during training with MEAN_STD, then unnormalized at inference). in LIBERO for example, these predicted actions are applied via a position controller, which matches the action space the model was trained on.
> 
> So if you’re deploying to IsaacSim, you’ll want to:
> 
> * unnormalize the actions before applying them.
> * feed them into your articulation controller as joint position targets (not raw deltas)
> * you can make sure/ base yourself that the control interface matches the LIBERO convention, otherwise the behavior may look erratic
> 
> At the end it depends on what data you want it trained on. And tip: make sure the observation key names matches what the policy expects! you can check libero docs here: https://huggingface.co/docs/lerobot/en/libero

Hi @jadechoghari , thanks for clarifications! I recently started to deploy SmolVLA in Isaac Sim with a simulated SO100. I haven't fine-tuned with any other data yet, just tried the zero-shot performance. The outcomes is different from what I expected, and I would appreicate to have some advices in potential reasons.

**The issue:**

The task is "pick the red sphere". The outcome is that the robot arm reached out firstly but always ended up stucking at the same position in the air. It always ended up in frontal direction, no matter where the red ball locates. The stuck position is shown in the [the final frame of the .gif in this readme](https://github.com/MyLovelyAxe/isaacsim_vla_ws).

**My setup:**

Isaac Sim ros2 nodes publish images and current joint states to a zmq socket;
SmolVLA subscribes to the observation socket for input, then returns absolute positions to an action zmq socket;
Isaac Sim gets actions from zmq socket and executes on the robot arm;

I have already unnormalized the input joint states and returned target joint states based on the simulated SO100 joint limits, [the script](https://github.com/MyLovelyAxe/lerobot/blob/bf8ffea66dbb61e11b8c359a05f078529ef657f8/examples/tutorial/smolvla/smolvla_zmq.py#L102) is here (I hope the unnormalization is done correctly).

I mainly would like to know whether this stucking behavior is expected for zero-shot performance, or any other potential directions to debug. Thank you very much for any suggestions!

**Comment by aditya1709** [comment]
@MyLovelyAxe One another thing I can think for this not working is the real to sim difference. Most of the SmolVLA models are trained on real data and they dont generalize well enough to the IsaacSim environment?

**Comment by MyLovelyAxe** [comment]
> [@MyLovelyAxe](https://github.com/MyLovelyAxe) One another thing I can think for this not working is the real to sim difference. Most of the SmolVLA models are trained on real data and they dont generalize well enough to the IsaacSim environment?

So if I fine-tune SmolVLA with synthetic dataset just generated using Isaac Sim, can the base model (i.e. the lerobot/smolvla_base) be able to improve? But maybe also will overfit if it is just fine-tuned on one single task like "pick up the ball"? @jadechoghari

**Comment by MyLovelyAxe** [comment]
@aditya1709  btw I have managed to make the returned absolute target joint states from SmolVLA work with the Articulation Controller node in Isaac Sim, by publishing the returned joint states value to topic `/joint_command`, and Articulation Controller will execute that in real-time. Here is the [.usd](https://github.com/MyLovelyAxe/isaacsim_vla_ws/blob/robot/so100/config/vla_so101_2cam.usd) for this setup.

**Comment by JulioPlaced** [comment]
Any updates on this? Did you manage to train SmolVLA in Isaac, @MyLovelyAxe ?
